pkg/machine/ocipull/ociartifact.go

```diff
@@ -0,0 +1,257 @@
+			return err
+		}
+		// grab the artifact disk out of the cache and lay
+		// it into our local cache in the format of
+		// hash + disktype + compression
+		//
+		// in cache it will be used until it is "outdated"
+		//
+		// i.e. 91d1e51...d28974.qcow2.xz
+		if err := o.unpack(artifactDigest); err != nil {
+			return err
+		}
+	} else {
+		logrus.Debugf("cached image exists and is latest: %s", cachedImagePath.GetPath())
+		o.cachedCompressedDiskPath = cachedImagePath
+	}
+	return o.decompress()
+}
+
+func (o *OCIArtifactDisk) getDestArtifact() (types.ImageReference, digest.Digest, error) {
+	imgRef, err := alltransports.ParseImageName(o.imageEndpoint)
+	if err != nil {
+		return nil, "", err
+	}
+	sysCtx := &types.SystemContext{
+		DockerInsecureSkipTLSVerify: types.NewOptionalBool(!o.pullOptions.TLSVerify),
+	}
+	imgSrc, err := imgRef.NewImageSource(o.ctx, sysCtx)
+	if err != nil {
+		return nil, "", err
+	}
+
+	defer func() {
+		if err := imgSrc.Close(); err != nil {
+			logrus.Error(err)
+		}
+	}()
+
+	diskArtifactDigest, err := GetDiskArtifactReference(o.ctx, imgSrc, o.diskArtifactOpts)
+	if err != nil {
+		return nil, "", err
+	}
+	// create a ref now and return
+	named := imgRef.DockerReference()
+	digestedRef, err := reference.WithDigest(reference.TrimNamed(named), diskArtifactDigest)
+	if err != nil {
+		return nil, "", err
+	}
+
+	// Get and "store" the original filename the disk artifact had
+	originalFileName, err := getOriginalFileName(o.ctx, imgSrc, diskArtifactDigest)
+	if err != nil {
+		return nil, "", err
+	}
+	o.diskArtifactFileName = originalFileName
+
+	newRef, err := docker.NewReference(digestedRef)
+	if err != nil {
+		return nil, "", err
+	}
+	return newRef, diskArtifactDigest, err
+}
+
+func (o *OCIArtifactDisk) pull(destRef types.ImageReference, artifactDigest digest.Digest) error {
+	destFileName := artifactDigest.Encoded()
+	destFile, err := o.dirs.ImageCacheDir.AppendToNewVMFile(destFileName, nil)
+	if err != nil {
+		return err
+	}
+	return Pull(o.ctx, destRef, destFile, o.pullOptions)
+}
+
+func (o *OCIArtifactDisk) unpack(diskArtifactHash digest.Digest) error {
+	finalSuffix := extractKindAndCompression(o.diskArtifactFileName)
+	blobDir, err := o.dirs.ImageCacheDir.AppendToNewVMFile(diskArtifactHash.Encoded(), nil)
+	if err != nil {
+		return err
+	}
+	cachedCompressedPath, err := o.dirs.ImageCacheDir.AppendToNewVMFile(diskArtifactHash.Encoded()+finalSuffix, nil)
+	if err != nil {
+		return err
+	}
+
+	o.cachedCompressedDiskPath = cachedCompressedPath
+
+	blobInfo, err := GetLocalBlob(o.ctx, blobDir.GetPath())
+	if err != nil {
+		return fmt.Errorf("unable to get local manifest for %s: %q", blobDir.GetPath(), err)
+	}
+
+	diskBlobPath := filepath.Join(blobDir.GetPath(), "blobs", "sha256", blobInfo.Digest.Encoded())
+
+	// Rename and move the hashed blob file to the cache dir
+	if err := os.Rename(diskBlobPath, cachedCompressedPath.GetPath()); err != nil {
```

Comment:
This could fail if the diskBlobPath is on a different disk then the  cachedCompressedPath.GetPath() and could be fixed by doing a copy/remove?

Reply:
lol, sure, you were the one that suggested i use rename to avoid a sparse clobering. maybe a move is safer?

Reply:
I was suggesting if Rename failed, then do Copy.